[llvm-lit] Add precommit test to verify current behavior of glob expansion in lit's internal shell

Harini0924 opened the pull request:
This patch introduces a precommit test to verify the current behavior of glob expansion in lit's internal shell. The motivation for this test stems from an issue encountered during the BOLT test suite when running with the lit internal shell using the command:

`LIT_USE_INTERNAL_SHELL=1 ninja check-bolt`

During execution, the following error was observed:
```
File "/usr/local/google/home/harinidonthula/llvm-project/llvm/utils/lit/lit/TestRunner.py", line 416, in executeBuiltinEcho
    stdout.write(encode(maybeUnescape(args[-1])))
TypeError: string argument expected, got 'GlobItem'
```
The `executeBuiltinEcho` function in the lit testing framework expects a string to be passed to `stdout.write`, but it received a `GlobItem` object instead. This precommit test is designed to check the current behavior where the glob pattern isn't correctly expanded, leading to this `TypeError`.

While this patch doesn't fix the issue, it helps in understanding and verifying the current behavior. The feedback I received from this [PR](https://github.com/llvm/llvm-project/pull/105925) suggests using `cmd.args = expand_glob_expressions(cmd.args, shenv.cwd)` to match the behavior of `executeBuiltinMkdir` and `executeBuiltinRm`, but it is recognized that the internal shell should ideally expand globs before calling any built-in command.

**Request for Feedback:**
I'm looking for feedback on how to improve this precommit test, specifically regarding the handling and expansion of glob patterns for commands like mkdir and rm within the internal shell. Currently, the args are expanded at the beginning of these functions, which should ensure proper glob expansion. However, I'd appreciate guidance on whether I should write additional tests to verify that mkdir and rm are handling glob expansions correctly.

If such tests are recommended, I would also appreciate advice on the best approach to implement them, considering the existing framework and the way glob expansion is expected to function in the internal shell. Should these tests confirm that the current implementation passes, or are there specific edge cases I should be aware of?

**Next Steps:**

In my follow-up PR, I plan to address the UNRESOLVED error by expanding the entire command, ensuring correct and consistent behavior across all commands. The current test checks for an unresolved issue with the glob expansion, specifically looking for a `TypeError` due to an unexpanded `GlobItem`. This will be updated to reflect the correct behavior once the issue is resolved.

This change is relevant for [[RFC] Enabling the Lit Internal Shell by Default](https://discourse.llvm.org/t/rfc-enabling-the-lit-internal-shell-by-default/80179/3)

Harini0924 committed:
[llvm-lit] Add shtest-glob.py test to verify glob pattern handling

This commit introduces a new test file, shtest-glob.py, to the lit testing framework. The test is designed to verify the handling of glob patterns within the internal shell, specifically checking the behavior of commands like echo when glob patterns are present. The test includes example files, example_file1.txt and example_file2.txt, which are used to validate whether the glob expansion is correctly performed or remains unexpanded as expected in certain cases. This addition helps ensure that glob pattern processing behaves consistently across different shell commands in lit.

Harini0924 committed:
Added the lit.cfg file

I forgot to add the lit.cfg file in my previous commit.

arichardson commented on llvm/utils/lit/tests/Inputs/shtest-glob/glob-echo.txt:
```diff
@@ -0,0 +1,2 @@
+## Tests glob pattern expansion by listing matching files.
+# RUN: echo %{inputs}/shtest-glob/example_file*.txt
```
Thanks, this is good. Can we also try testing a glob expansion passes to a non-builtin command? E.g. a local python script?

arichardson replied:
Could also do something like `touch %t1.txt %t2.txt` and then try `mkdir %t*.txt` and check that we get two file exists errors.

Harini0924 replied:
Thank you for the suggestions! I have a couple of follow-up questions to ensure I'm on the right track.

For testing glob expansion with a non-built-in command, would it be appropriate to use the `find` command? I was considering using `find` to list files that match a glob pattern. Do you think this would effectively demonstrate the glob expansion, or is there another command you would recommend?

Regarding the mkdir suggestion, I'm a bit confused about where this test should be placed. Should it go in the same file as the glob-echo test (`llvm/utils/lit/tests/Inputs/shtest-glob/glob-echo.txt`), or would it be better to create a new file specifically for mkdir, such as `llvm/utils/lit/tests/Inputs/shtest-glob/glob-mkdir.txt`? And then, in `llvm/utils/lit/tests/shtest-glob.py`, should this test be expected to fail, indicating the current behavior before implementing the fix?

arichardson replied:
I'm not too familiar with the structure of the lit internal test suite - I'd say keeping the tests in separate files matches the current layout more closely. Looking at the code mkdir handles globs already so if you just run something like `not mkdir %S/example_input%.input` you can verify the error output.

arichardson commented on llvm/utils/lit/tests/Inputs/shtest-glob/example_file1.txt:
```diff
@@ -0,0 +1,2 @@
+## This is the first example file used for testing glob pattern matching.
+RUN: This is the first example file.
```
```suggestion
```
This file should not contain any RUN: lines especially not ones that cannot actually be run. If you rename it to something like `example_file1.input` lit will not try to run it (since we have `config.suffixes = [".txt"]`).

arichardson commented on llvm/utils/lit/tests/Inputs/shtest-glob/example_file2.txt:
```diff
@@ -0,0 +1,2 @@
+## This is the second example file used for testing glob pattern matching.
+RUN: This is the second example file.
```
```suggestion
```
same here - also rename it to have a different suffix.

Harini0924 replied:
I wasn’t too sure when you suggested renaming the suffix. I assumed you meant renaming this file suffix to `.input` to match the other file, so I did that. However, if you intended for a different suffix, like `.data` or something else, I can make that change. Please let me know if you have a specific suffix in mind.

arichardson replied:
Yes that's exactly what I meant. The suffix doesn't matter we just want two files matching the glob.

arichardson commented on llvm/utils/lit/tests/Inputs/shtest-glob/glob-echo.txt:
```diff
@@ -0,0 +1,2 @@
+## Tests glob pattern expansion by listing matching files.
+# RUN: echo %{inputs}/shtest-glob/example_file*.txt
```
```suggestion
# RUN: echo %S/example_file*.input
```

Harini0924 replied:
done.

arichardson reviewed (changes requested):
(accidentally accepted)

Harini0924 committed:
[llvm-lit] Rename and update test files for glob pattern handling

Renamed `example_file1.txt` and `example_file2.txt` to `.input` files to avoid `RUN` lines in test files and ensure they are not executed as test scripts. Added `glob-mkdir.txt` to test `mkdir` with glob patterns, ensuring failure when directories exist.  Modified `shtest-glob.py` to include checks for both `echo` and `mkdir` command tests. Ensured that the `mkdir` command test is expected to fail, confirming proper error handling for existing paths.

Harini0924 commented:
I will address the code formatter issues later.

Harini0924 commented:
@arichardson I followed your suggestion about keeping the tests in separate files to align with the current structure of the lit internal test suite. I created a file `glob-mkdir.txt` that checks if the mkdir command fails as expected when attempting to create a directory that already exists (using `example_file`). This is reflected in the `shtest-glob.py file`, where the test is expected to fail. Is this what you had in mind? Should I add or change anything else?

ilovepi reviewed (commented):
LGTM, as long as @arichardson is satisfied.

arichardson commented on llvm/utils/lit/tests/shtest-glob.py:
```diff
@@ -0,0 +1,13 @@
+# CHECK: TypeError: string argument expected, got 'GlobItem'
+
+# CHECK: FAIL: shtest-glob :: glob-mkdir.txt ({{[^)]*}})
+# CHECK: # error: command failed with exit status: 1
```
I was surprised there is no more specific error here but it looks like the builtin mkdir does not print any output right now. Turns out it doesn't handle EEXISTS the same way as the real mkdir command. I can submit a patch for this once this commit lands.

Harini0924 replied:
A patch to address that would be great, and I’m looking forward to seeing it once this commit lands.

arichardson reviewed (approved):
Thanks, this looks good to me.

arichardson commented:
> Thanks, this looks good to me.

But please do fix the trailing whitespace before committing.

Harini0924 committed:
Fixed the trailing whitespace
